pkg/core/kustomize/kustomize.go

```diff
@@ -0,0 +1,141 @@
+package kustomize
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"sigs.k8s.io/kustomize/k8sdeps"
+	"sigs.k8s.io/kustomize/pkg/commands/build"
+	"sigs.k8s.io/kustomize/pkg/fs"
+	"time"
+)
+
+type Kustomizer interface {
+	// Applies the provided label to the provided remote or local resource definition
+	// Returns the customized resource contents
+	// Returns an error if
+	// - the URL scheme is not supported (only file, http and https are)
+	// - retrieving the content fails
+	// - applying the customization fails
+	ApplyLabel(resourceUri *url.URL, label *Label) ([]byte, error)
+}
+
+type Label struct {
+	Name  string
+	Value string
+}
+
+func (l *Label) AsMap() map[string]string {
+	return map[string]string{l.Name: l.Value}
+}
+
+type kustomizer struct {
+	fakeDir    string
+	fs         fs.FileSystem
+	httpClient *http.Client
+}
+
+func MakeKustomizer(timeout time.Duration) Kustomizer {
+	return &kustomizer{
+		fs:      fs.MakeFakeFS(),
+		fakeDir: "/",
+		httpClient: &http.Client{
+			Timeout: timeout,
+		},
+	}
+}
+
+func (kust *kustomizer) ApplyLabel(resourceUri *url.URL, label *Label) ([]byte, error) {
+	resourcePath, err := kust.writeResourceFile(resourceUri)
+	if err != nil {
+		return nil, err
+	}
+	err = kust.writeKustomizationFile(resourcePath, label)
+	if err != nil {
+		return nil, err
+	}
+	return kust.runBuild()
+}
+
+func (kust *kustomizer) writeResourceFile(resourceUri *url.URL) (string, error) {
+	resourceContents, err := kust.fetch(resourceUri)
+	if err != nil {
+		return "", err
+	}
+	resourcePath := "resource.yaml"
+	err = kust.fs.WriteFile(kust.fakeDir+resourcePath, []byte(resourceContents))
```

Comment:
What stops resource files for distinct resources colliding?

Reply:
there is only one resource file per `ApplyLabel` call, I don't think there can be be collisions in that case.

Reply:
one thing I could do at the end of `ApplyLabel` is have a deferred function call that resets the `map[string]*FakeFile`, which underlies the in-memory filesystem implementation that is used here.

Reply:
> there is only one resource file per `ApplyLabel` call, I don't think there can be be collisions in that case.

Fair enough, thanks.

Strictly speaking (and I know the current caller doesn't do this, but) if multiple calls to `ApplyLabel` were made concurrently, then collisions would occur. If this bothers you, you could add a godoc comment to `ApplyLabel` saying it is not suitable for concurrent invocation.